File: src/446_Arithmetic_Slices_II_-_Subsequence.rs
// 446_Arithmetic_Slices_II_-_Subsequence
// https://leetcode.cn/problems/arithmetic-slices-ii-subsequence/description/

use std::collections::HashMap;

impl Solution {
    pub fn number_of_arithmetic_slices(nums: Vec<i32>) -> i32 {
        //将nums中的元素转换为i64类型
        let nums: Vec<i64> = nums.iter().map(|&x| x as i64).collect();
        let n = nums.len();
        //使用HashMap存储子序列的差值和个数
        let mut dp = HashMap::new();
        let mut ans = 0;

        //枚举i和j，计算以i结尾、差值为diff的等差数列个数
        for i in 0..n {
            for j in 0..i {
                let diff = nums[i] - nums[j];
                //获取以j结尾、差值为diff的等差数列个数，若不存在则设为0
                let count = dp.get(&(j, diff)).unwrap_or(&0).clone();
                //将以i结尾、差值为diff的等差数列个数加上以j结尾、差值为diff的等差数列个数，并更新HashMap
                dp.entry((i, diff)).and_modify(|x| *x += count + 1).or_insert(count + 1);
                //累加以j结尾、差值为diff的等差数列个数
                ans += count;
            }
        }
        //返回等差数列个数
        ans as i32
    }
}

File: src/446_Arithmetic_Slices_II_-_Subsequence.py
# // 446_Arithmetic_Slices_II_-_Subsequence
# // https://leetcode.cn/problems/arithmetic-slices-ii-subsequence/description/

# 给你一个整数数组 nums ，返回 nums 中所有 等差子序列 的数目。

# 如果一个序列中 至少有三个元素 ，并且任意两个相邻元素之差相同，则称该序列为等差序列。

# 例如，[1, 3, 5, 7, 9]、[7, 7, 7, 7] 和 [3, -1, -5, -9] 都是等差序列。
# 再例如，[1, 1, 2, 5, 7] 不是等差序列。
# 数组中的子序列是从数组中删除一些元素（也可能不删除）得到的一个序列。

# 例如，[2,5,10] 是 [1,2,1,2,4,1,5,10] 的一个子序列。
# 题目数据保证答案是一个 32-bit 整数。

from typing import List
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        ans = 0
        f = [defaultdict(int) for _ in nums]
        for i, x in enumerate(nums):
            for j in range(i):
                d = x - nums[j]
                cnt = f[j][d]
                ans += cnt
                f[i][d] += cnt + 1
        return ans